feat(metrics_indexer): Add rate limits functionality to indexer, take 2 [INGEST-1380]

untitaker opened the pull request:
This reverts commit a65c78a.

Reverts https://github.com/getsentry/sentry/pull/36524
Reapplies https://github.com/getsentry/sentry/pull/36263 + typing fixes

## Split up in the following PRs

Merge in this order (it should be safe to deploy after each merge)

- [x] https://github.com/getsentry/sentry/pull/37168
- [x] https://github.com/getsentry/sentry/pull/37170
- [x] https://github.com/getsentry/sentry/pull/37171
- [x] https://github.com/getsentry/sentry/pull/37173
- [ ] https://github.com/getsentry/sentry/pull/37174

untitaker committed:
Revert "Revert "feat(metrics_indexer): Add rate limits functionality to indexer [INGEST-1380] (#36263)" (#36524)"

This reverts commit a65c78a92426841112de15e2f0d9168c57d8b352.

untitaker committed:
fix typing of indexer service and fix resulting typing errors

untitaker committed:
remove assertions just to make tests pass

untitaker committed:
fix test crashes

fpacifici reviewed (commented):
Since this is the second time we deplyo a change to the processing phase of the indexer and we miss a bug in testing, I believe we have two problems:
- the abstraction in the indexer is not right so it makes it hard to make changes with confidence.
- we lack unit tests.

I have an idea to solve both, and I think it would be worth to write another PR before this one to implement the idea. It should not be too big:
Basically what happens [here](https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/consumers/indexer/processing.py#L63-L239) is:
- for each batch
- extract all the strings to resolve from the batch
- talk to the indexer to resolve those strings
- validate data is ok
- fit the integers into the batch 
- compose the kafka messages.

What if we had a smarter `Batch` abstraction that takes care of the transformation:
```
class MessageBatch: # now this is jsut a List
     def __init__(List[Messages]):
         ....

     def extract_strinags(self) -> SomeStringsBag like `KeyCollection` 

     def map(self, KeyResult) -> MessageBatch[str]
          """
          This produces a new batch (actually modifying the strings in place) 
          with the mapped strings
          """
```

So the processor coordinates the work. It asks the batch to provide the strings. It validates them, then taks to the indexer to resolve them and record. Then it talks again to the batch to produce a new output batch with the results. The batch data structure knows the message payload, it can extract strings and map them back.

Now the batch data structure is stateless and can be unit tested to death.
What do you think ?

untitaker commented:
@fpacifici We have integration tests that ended up not testing anything because they are using the mock indexer (which doesn't have rate limits) together with mocked/stubbed fetch results coming from the rate limiter, and those went out of sync with the actual implementation once I added FetchTypeExt in the final round of reviews.

We want to be able to test message processing and rate limiting together, i.e. their interaction, because that's the boundary where things broke. At the same time, if we want to test this integration to death we ideally find a way to do it without hitting the database in order to keep the testsuite fast. So this suggests to me that we should figure out a way to detach rate limiting (and maybe caching as well) from the postgres_v2 backend, so it can be tested together with message processing without hitting the DB.

So let's say we build this Batch abstraction, and if I understand you right it would not handle rate limiting, and the responsibilities of `indexer.bulk_record` would stay the same. Then we can unittest that to death, but as usual those unittests will need to simulate whatever can be returned from the indexer (and that can go out of sync with the actual implementation again), or they hit the real postgres indexer (in which case the tests will be very slow). That's fundamentally the same problem we're dealing with already.

lmk what you think

fpacifici commented:
I think the two approaches are not incompatible. 
You are right we need a better way to test what comes from the indexer and rate limiter.
Still testing the processing logic in isolation is not supposed to be alternative to that. 
While this time the issue was related with the data returned by the indexer, last time it was pure processing. 

But I would be fine if you do not want to break it apart within this task.

fpacifici commented:
Another idea to make it harder to break and easier to test 

Why don't we return a properly structured type here, the current state feels like a case of Primitive Obsession  :
https://github.com/getsentry/sentry/pull/36531/files#diff-4d832960b67c8c52695fce355ad53704c4c0bab03ef0fdd5c487817d383094edR162


If that returned a `Mapping[str, Metadata]` where `Metadata` is a dataclass or a named tuple with properties instead of a bare tuple, the type checker should have noticed the issue. We would not be able to do tuple unpacking. 
That `KeyResult` originally returned `Mapping[str, int]` so fairly harmless. Then we added metadata, then FetchType and then FetchTypeExt. Now it is likely abusing that original Mapping.

untitaker committed:
introduce Metadata struct

untitaker commented:
@fpacifici had to do a double-take on your first post, if I understood it right then it's actually really simple and https://github.com/getsentry/sentry/pull/36669 is basically it

untitaker committed:
Merge remote-tracking branch 'origin/master' into feat/indexer-rate-limits-2

untitaker committed:
fix typing everywhere, remove unused mocks, and reintroduce changes lost during merge

untitaker commented on src/sentry/sentry_metrics/consumers/indexer/processing.py:
```diff
@@ -1,29 +1,16 @@
 logger = logging.getLogger(__name__)
 
 
-@functools.lru_cache(maxsize=10)
```
I got rid of those because otherwise the return type of `indexer` is `Any` in `process_messages`, and that would infect return values of `indexer.bulk_record`, and `get_fetch_metadata` (where i shipped a bug/crash last time)

untitaker commented on tests/sentry/sentry_metrics/test_multiprocess_steps.py:
```diff
@@ -260,8 +260,7 @@ def __translated_payload(
     return payload
 
 
-@patch("sentry.sentry_metrics.consumers.indexer.processing.get_indexer", return_value=MockIndexer())
```
since https://github.com/getsentry/sentry/pull/36397 this is no longer necessary. the global indexer is mocked and cleared on every testrun because of the `sentry_metrics` marker, because this exact decorator was scattered across hundreds of product/api tests as well it seemed like a good default

untitaker committed:
revert fork-unsafe changes

fpacifici commented on src/sentry/runner/commands/run.py:
```diff
@@ -552,7 +552,7 @@ def ingest_consumer(consumer_types, all_consumer_types, **options):
 @click.option("--output-block-size", type=int, default=DEFAULT_BLOCK_SIZE)
 @click.option("--factory-name", default="default")
-@click.option("--ingest-profile")
+@click.option("--ingest-profile", required=True)
```
Could you please double check that we are passing the ingest profile in prod for Release health before merging ?

fpacifici commented on src/sentry/sentry_metrics/consumers/indexer/batch.py:
```diff
@@ -138,8 +138,8 @@ def extract_strings(self) -> Tuple[Mapping[int, Set[str]], Set[str]]:
-        mapping: Mapping[int, Mapping[str, int]],
-        bulk_record_meta: Mapping[str, Tuple[int, FetchType]],
+        mapping: Mapping[int, Mapping[str, Optional[int]]],
+        bulk_record_meta: Mapping[str, Metadata],
```
Not sure I get why this is not broken by org. 
The same string can be in two orgs and we could be able to write it in one and not in the other.

untitaker replied:
it's a long-standing bug. I pointed it out in sns channel a few months ago. we should fix it, thought somebody else was going to do it. will do now

fpacifici commented on src/sentry/sentry_metrics/consumers/indexer/batch.py:
```diff
@@ -161,26 +161,81 @@ def reconstruct_messages(
                 logger.error("process_messages.key_error", extra={"tags": tags}, exc_info=True)
                 continue
 
+            if exceeded_org_quotas or exceeded_global_quotas:
```
Please write a datadog metric here as well. The error is likely going to be dropped (we do not need sentry errors when a customer floods us) though the metric will still be useful

fpacifici commented on src/sentry/sentry_metrics/consumers/indexer/batch.py:
```diff
@@ -161,26 +161,81 @@ def reconstruct_messages(
             new_payload_value["tags"] = new_tags
-            new_payload_value["metric_id"] = mapping[org_id][metric_name]
+            new_payload_value["metric_id"] = numeric_metric_id = mapping[org_id][metric_name]
+            if numeric_metric_id is None:
```
Same as above, let's have a metric

fpacifici commented on src/sentry/sentry_metrics/indexer/ratelimiters.py:
```diff
@@ -0,0 +1,151 @@
+        raise ValueError(use_case_id)
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quota_requests")
```
Why do we need to collect metrics out of this method and the settings one? They are supposed to have a minimal impact, as they only build data structures from the input provided and have no external dependencies.

untitaker replied:
yes ideally they would, but as we saw with reconstruct-messages, python code can have some significant overhead

fpacifici commented on src/sentry/sentry_metrics/indexer/ratelimiters.py:
```diff
@@ -0,0 +1,151 @@
+    return org_ids, requests
+
+
+RateLimitState = Tuple[UseCaseKey, Sequence[RequestedQuota], Sequence[GrantedQuota], Timestamp]
```
Please make this a NamedTuple or it will be impossible to manage.

untitaker replied:
made it dataclass, done

fpacifici commented on src/sentry/sentry_metrics/indexer/ratelimiters.py:
```diff
@@ -0,0 +1,151 @@
+    @metrics.wraps("sentry_metrics.indexer.check_write_limits")
+    def check_write_limits(
+        self, use_case_id: UseCaseKey, keys: KeyCollection
+    ) -> Tuple[RateLimitState, KeyCollection, Sequence[Tuple[KeyResult, FetchType, FetchTypeExt]]]:
```
A tuple containing a Tuple and Sequence of tuples should have a proper structured type. Otherwise it will be impossible to maintain.

untitaker replied:
I moved the innermost tuple to a dataclass, lmk if that is sufficient

fpacifici commented on src/sentry/sentry_metrics/indexer/ratelimiters.py:
```diff
@@ -0,0 +1,151 @@
+        org_ids, requests = _construct_quota_requests(use_case_id, keys)
+        timestamp, grants = self._get_rate_limiter(use_case_id).check_within_quotas(requests)
+
+        granted_key_collection = dict(keys.mapping)
```
Why do you need to make a copy of the mapping ?

Also why is this called `granted_key_collection` ? This is the original mapping. It includes both allowed and rejected keys.

untitaker replied:
it doesn't. in the last line of the for-loop we write back the list of strings that fit into the quota.

fpacifici commented on src/sentry/sentry_metrics/indexer/ratelimiters.py:
```diff
@@ -0,0 +1,151 @@
+@metrics.wraps("sentry_metrics.indexer.construct_quota_requests")
+def _construct_quota_requests(
+    use_case_id: UseCaseKey, keys: KeyCollection
+) -> Tuple[Sequence[OrgId], Sequence[RequestedQuota]]:
```
Why not returning a mapping? It seems it would be a lot more readable.

untitaker replied:
this was already a topic last time I attempted to merge this patchset into master, I wanted to avoid unnecessary type conversions between dict and list.

fpacifici commented on src/sentry/sentry_metrics/indexer/ratelimiters.py:
```diff
@@ -0,0 +1,151 @@
+
+            granted_key_collection[org_id] = allowed_strings
+
+        state = use_case_id, requests, grants, timestamp
+        return state, KeyCollection(granted_key_collection), dropped_key_results
```
Why do we need to return the use case id and the requests ?

untitaker replied:
this `state` is an "opaque" value that the indexer shouldn't poke at and pass to `apply_write_limits`. I thought this would be the best way to ensure we read and write with the same usecase.

fpacifici commented on src/sentry/sentry_metrics/indexer/ratelimiters.py:
```diff
@@ -0,0 +1,151 @@
+RateLimitState = Tuple[UseCaseKey, Sequence[RequestedQuota], Sequence[GrantedQuota], Timestamp]
+
+
+class WritesLimiter:
```
Could you please add some unit tests for this class? It is complex enough that we can mess up corner cases.

fpacifici commented on src/sentry/sentry_metrics/consumers/indexer/batch.py:
```diff
@@ -138,8 +138,8 @@ def extract_strings(self) -> Tuple[Mapping[int, Set[str]], Set[str]]:
         self,
-        mapping: Mapping[int, Mapping[str, int]],
-        bulk_record_meta: Mapping[str, Tuple[int, FetchType]],
+        mapping: Mapping[int, Mapping[str, Optional[int]]],
```
Please let's have a unit test for this method. 
There are at least 5 code paths to cover:
- all good, all strings resolved
- metric id rate limited 
- some tag keys rate limited
- some tag values rate limited
- the same string is successful in an organization and not in another (two different messages). 

While the end to end test for the `process` method is good and useful, you would not test each and every one of those code paths above in an end to end test. This is one of the advantage of breaking this logic out of the massive process method we had before. Now we can test each case separately without going crazy.

fpacifici reviewed (commented):
Please see my comments in line.

Also I would consider splitting this PR. There is a lot going on. 
What about introducing the rate limiter without taking the results into account first ? Which means make the changes to hte sliding window rate limtier, add the `ratelimiters.py` class and wire it into the indexer, but still write all the strings on the DB and do not change anything in the reconstruct ? 
This way we would reduce the risk by introducing the rate limiter logic without impacting the messages in any way.

untitaker committed:
add unittests, fix metadata bugs

untitaker committed:
add dd metrics

untitaker committed:
make ratelimitstate a dataclass

untitaker committed:
fix tests around meta bugfix

untitaker committed:
add rate limiters tests

untitaker commented:
@fpacifici thanks for the thorough review and for keeping coding standards high. I think I have addressed all review comments.

I can split up the PR. Should I do this before or after you review the recent changes?

untitaker committed:
add missing type annotations

fpacifici commented:
> I can split up the PR. Should I do this before or after you review the recent changes?

Sorry I did not see this comment before. I think it is good to split. No need for another round. Though I will try to do another review tonight before you will wake up. 
If you do not see it by the time you start working, please go ahead.

untitaker commented:
I split up the PR into five different ones and edited the PR description to reflect in which order they need to be merged. After that I'll rebase this PR and hopefully the diff empty, if not I guess I'll just merge it.
